pandapower/test/shortcircuit/test_sgen.py

```diff
@@ -89,13 +89,13 @@ def test_min_branch_results_1(three_bus_example):
     assert np.allclose(net.res_line_sc.ip_ka.values, np.array([0.01781447, 0.]))
     assert np.allclose(net.res_line_sc.ith_ka.values, np.array([0.01265116, 0.]))
 
-def test_min_branch_results_2(big_sgen_three_bus_example):
+def test_min_branch_results_big_sgen(big_sgen_three_bus_example):
     net = big_sgen_three_bus_example
     sc.calc_sc(net, case="min", ip=True, ith=True, branch_results=True)
     assert np.allclose(net.res_bus_sc.ikss_ka.values, np.array([1.67956442, 1.65864191, 1.62941387]))
     assert np.allclose(net.res_line_sc.ikss_ka.values, np.array([0.36974055, 0.]))
-    assert np.allclose(net.res_line_sc.ip_ka.values, np.array([0.69978051,   0.]))
-    assert np.allclose(net.res_line_sc.ith_ka.values, np.array([0.37135563,  0.]))
+    assert np.allclose(net.res_line_sc.ip_ka.values, np.array([0.69687302,   0.]))
+    assert np.allclose(net.res_line_sc.ith_ka.values, np.array([0.37133258,  0.]))
 

```

Comment:
You changed the expected results here - where did you take these values from?

Reply:
I solved the circuit analytically, the calculation are stored in a jupyter notebook that I can share if you for a cross check if you want.

Reply:
Here the solution:

```
def three_bus_big_sgen_example():
    #
    #      Bus0    Bus1     Bus2   
    #      _  line0 _  line1 _ 
    # EG--|_|------|_|------|_|
    #              |              
    #             SG
    #
    net = pp.create_empty_network()
    b1 = pp.create_bus(net, 110)
    b2 = pp.create_bus(net, 110)
    b3 = pp.create_bus(net, 110)

    pp.create_ext_grid(net, b1, s_sc_max_mva=100., s_sc_min_mva=80., rx_min=0.4, rx_max=0.4)
    pp.create_line(net, b1, b2, std_type="305-AL1/39-ST1A 110.0" , length_km=20.)
    pp.create_line(net, b2, b3, std_type="N2XS(FL)2Y 1x185 RM/35 64/110 kV" , length_km=15.)
    net.line["endtemp_degree"] = 80

    pp.create_sgen(net, b2, sn_mva=200., p_mw=0, k=1.2)
    return net

net = three_bus_big_sgen_example()

# cmin
c = 1.

# Base quantities
Pbase = net.sn_mva
Ibase = Pbase / (net.bus.vn_kv.values * np.sqrt(3)) 
Zbase = (net.bus.vn_kv.values**2) / Pbase

#ext_grid impedance
s_sc_mva = 80.
rx = 0.4
z_ext_grid_abs = c / s_sc_mva 
x_ext_grid = z_ext_grid_abs / np.sqrt(1+rx**2) 
r_ext_grid = x_ext_grid * rx 
z_ext_grid = (r_ext_grid + x_ext_grid*1j) * Pbase

#line impedances
K = 1. + 0.004 * (80. - 20.)
z_line0 =(net.line.r_ohm_per_km.values[0]*K + net.line.x_ohm_per_km.values[0]*1j)*net.line.length_km[0] / Zbase[1]
z_line1 =(net.line.r_ohm_per_km.values[1]*K + net.line.x_ohm_per_km.values[1]*1j)*net.line.length_km[1] / Zbase[2]

#current source current
i_gen = net.sgen.k * net.sgen.sn_mva / Pbase * Ibase[0] 

# ikssa bus
i_bus0 = np.abs(c / z_ext_grid * Ibase[0]) + np.abs(i_gen)
i_bus1 = np.abs(c / (z_ext_grid + z_line0) * Ibase[1]) + np.abs(i_gen)
i_bus2 = np.abs(c / (z_ext_grid + z_line0 + z_line1) * Ibase[2]) +\
         np.abs(i_gen * (z_ext_grid + z_line0) / (z_ext_grid + z_line0 + z_line1))
i_bus_vec = np.array([i_bus0, i_bus1, i_bus2])

# line currents network contrib
i_line00_net = 0.
i_line01_net = np.abs(c / (z_ext_grid + z_line0) * Ibase[1])
i_line02_net = np.abs(c / (z_ext_grid + z_line0 + z_line1) * Ibase[2])

i_line10_net = 0.
i_line11_net = 0.
i_line12_net = np.abs(c / (z_ext_grid + z_line0 + z_line1) * Ibase[2])

i_line_mat_net = np.array([[i_line00_net, i_line01_net, i_line02_net],
                           [i_line10_net, i_line11_net, i_line12_net]])

#array([[0.        , 0.3989686 , 0.39193811],
#       [0.        , 0.        , 0.39193811]])

# line currents current source contrib
i_line00_gen = np.abs(i_gen)
i_line01_gen = 0.
i_line02_gen =  - (np.abs(i_gen) - np.abs(i_gen * (z_ext_grid + z_line0) / (z_ext_grid + z_line0  + z_line1)))

i_line10_gen = 0.
i_line11_gen = 0.
i_line12_gen = np.abs(i_gen * (z_ext_grid + z_line0) / (z_ext_grid + z_line0 + z_line1))

i_line_mat_gen = np.array([[i_line00_gen, i_line01_gen, i_line02_gen],
                           [i_line10_gen, i_line11_gen, i_line12_gen]])

#array([[ 1.25967331,  0.        , -0.02219755],
#       [ 0.        ,  0.        ,  1.23747576]])

# Line current total
i_line_mat = i_line_mat_net + i_line_mat_gen

#array([[1.25967331, 0.3989686 , 0.36974055],
#       [0.        , 0.        , 1.62941387]])

# I peak
ip_line_mat = np.sqrt(2) * (i_line_mat_net * bus3._ppc["bus"][:, KAPPA] + i_line_mat_gen)

#array([[1.78144709, 0.74438751, 0.69687302],
#       [0.        , 0.        , 2.47832011]])

# I thermal
ith_line_mat = i_line_mat * np.sqrt(1 + bus3._ppc["bus"][:, M])

#array([[1.26511638, 0.40071219, 0.37133258],
#       [0.        , 0.        , 1.63642978]])
```

Reply:
Thanks for following up! We usually like to validate our results with other software instead of own calculation, to get an external reference, to avoid that there is a bug in the reference calculation.

All results that are currently in these tests are copied from PowerFactory, the test grids are in test/test_files/sc_test.pfd. I would like to test these changes agains PowerFactory or another commercial software to be sure that everything is correct.

Reply:
Sure, go ahead. I would have done it myself but I don't have any other commercial software to compare with.

Reply:
So I am trying to reproduce the motor examples, but without any success until now. I am getting completely different results with PowerFactory for the motor short-circuit current. You are definitely right that there is something wrong there, the only test that existed only passed because there were no assert statements :/ 

How did you create the test results for the motor test cases?

Reply:
Max shortcircuit:
```
def motor_three_bus_net():
    # b1    b2     b3
    # o-----o------o
    # |     |
    #eg     G  
    net = pp.create_empty_network(sn_mva=1.)
    b1 = pp.create_bus(net, vn_kv=0.4)
    b2 = pp.create_bus(net, vn_kv=0.4)
    b3 = pp.create_bus(net, vn_kv=0.4)
    
    pp.create_sgen(net, b2, p_mw=0.11, sn_mva=5., type="motor", k=7, rx=0.6, current_source=False)
    pp.create_line_from_parameters(net, from_bus=b1, to_bus=b2, length_km=1., r_ohm_per_km=0.3211,
                                   x_ohm_per_km=0.06911504, c_nf_per_km=0, max_i_ka=1)
    pp.create_line_from_parameters(net, from_bus=b2, to_bus=b3, length_km=1., r_ohm_per_km=0.3211,
                                   x_ohm_per_km=0.06911504, c_nf_per_km=0, max_i_ka=1)

    net.line["endtemp_degree"] = 165
    pp.create_ext_grid(net, b1, s_sc_max_mva=5., s_sc_min_mva=1., rx_min=0.4, rx_max=0.4)
    
    return net

# Constants
c = 1.1
net = motor_three_bus_net()
Pbase = net.sn_mva
Ibase = Pbase / (net.bus.vn_kv.values * np.sqrt(3))
Zbase = (net.bus.vn_kv.values**2) / Pbase

# ext_grid impedance
s_sc_max_mva=5.
rx_max=0.4
z_ext_grid_abs = c / s_sc_max_mva 
x_ext_grid = z_ext_grid_abs / np.sqrt(1+rx_max**2) 
r_ext_grid = x_ext_grid * rx_max 
z_ext_grid = (r_ext_grid + x_ext_grid*1j) * Pbase

# line impedances
z_line1 =(0.3211 + 0.06911504j)*1.0 / Zbase[1]
z_line2 =(0.3211 + 0.06911504j)*1.0 / Zbase[2]

#motor impedance
k = 7.
sn_gen = 5.
rx = 0.6
vn_kv = net.bus.loc[net.sgen.bus.values[0]].vn_kv
z_motor_abs = 1./k * (vn_kv**2) / sn_gen * 1000. # ohm
x_motor = z_motor_abs/np.sqrt(1 + rx**2)
r_motor = rx * x_motor
z_motor = (r_motor + x_motor*1j)/Zbase[1]

#bus currents
i_b1 = c / (z_motor + z_line1) + c / z_ext_grid
i_b1 = np.abs(i_b1 * Ibase[0])

i_b2 = c / (z_ext_grid + z_line1) + c / z_motor
i_b2 = np.abs(i_b2 * Ibase[1])

i_b3 = c / (z_line2 + z_motor * (z_ext_grid + z_line1) / (z_motor + z_ext_grid + z_line1))
i_b3 = np.abs(i_b3 * Ibase[2])

i_bus = np.array([i_b1, i_b2, i_b3])

# array([7.26862889, 0.76933315, 0.38652958])

# Branch currents
i_line_00 = np.abs(c / (z_motor + z_line1)) * Ibase[0]
i_line_01 = np.abs(c / (z_ext_grid + z_line1)) * Ibase[1]
i_line_02 = c / (z_ext_grid + z_line1 + z_motor * z_line2 / (z_motor + z_line2))
i_gen_02 = c / (z_motor + (z_ext_grid + z_line1) * z_line2 / (z_ext_grid + z_line1 + z_line2))
i_line_02 -= i_gen_02 * z_line2 / (z_ext_grid + z_line1 + z_line2)
i_line_02 = np.abs(i_line_02) * Ibase[2]

i_line_10 = 0.
i_line_11 = 0.
i_ext_12 = c / (z_ext_grid + z_line1 + z_motor * z_line2 / (z_motor + z_line2))
i_gen_12  = c / (z_motor + (z_ext_grid + z_line1) * z_line2 / (z_ext_grid + z_line1 + z_line2))
i_line_12 = i_ext_12 * z_motor / (z_motor + z_line2)
i_line_12 += i_gen_12 * (z_ext_grid + z_line1) / (z_ext_grid + z_line1 + z_line2)
i_line_12 = np.abs(i_line_12) * Ibase[2]


i_line_mat = np.array([[i_line_00, i_line_01, i_line_02],
                       [i_line_10, i_line_11, i_line_12]]) 


# array([[0.05290295, 0.72719542, 0.36535867],
#        [0.        , 0.        , 0.38652958]])

net = motor_three_bus_net()
sc.calc_sc(net, case='max', branch_results=True)
```

Min shortcircuit:

```
def motor_three_bus_net():
    #b1    b2     b3
    # o-----o------o
    # |     |
    # Eg    G  
    net = pp.create_empty_network(sn_mva=1.)
    b1 = pp.create_bus(net, vn_kv=0.4)
    b2 = pp.create_bus(net, vn_kv=0.4)
    b3 = pp.create_bus(net, vn_kv=0.4)
    
    pp.create_sgen(net, b2, p_mw=0.11, sn_mva=5., type="motor", k=7, rx=0.6, current_source=False)
    pp.create_line_from_parameters(net, from_bus=b1, to_bus=b2, length_km=1., r_ohm_per_km=0.3211,
                                   x_ohm_per_km=0.06911504, c_nf_per_km=0, max_i_ka=0.3)
    pp.create_line_from_parameters(net, from_bus=b2, to_bus=b3, length_km=1., r_ohm_per_km=0.3211,
                                   x_ohm_per_km=0.06911504, c_nf_per_km=0, max_i_ka=0.3)

    net.line["endtemp_degree"] = 165
    pp.create_ext_grid(net, b1, s_sc_max_mva=5., s_sc_min_mva=1., rx_min=0.4, rx_max=0.4)
    #pp.create_switch(net, b3, b1, et="b")
    return net

# Constants
c = 0.95
net = motor_three_bus_net()
Pbase = net.sn_mva
Ibase = Pbase / (net.bus.vn_kv.values * np.sqrt(3))
Zbase = (net.bus.vn_kv.values**2) / Pbase

# ext_grid impedance
s_sc_min_mva=1.
rx_min=0.4
z_ext_grid_abs = c / s_sc_min_mva 
x_ext_grid = z_ext_grid_abs / np.sqrt(1+rx_min**2) 
r_ext_grid = x_ext_grid * rx_min 
z_ext_grid = (r_ext_grid + x_ext_grid*1j) * Pbase

# line impedances
kl = 1 + 0.004 * (165 -20)
z_line1 =(0.3211 * kl + 0.06911504j)*1.0/Zbase[1]
z_line2 =(0.3211 * kl + 0.06911504j)*1.0/Zbase[2]

#motor impedance
k = 7.
sn_gen = 5.
rx = 0.6
vn_kv = net.bus.loc[net.sgen.bus.values[0]].vn_kv
z_motor_abs = 1./ k * (vn_kv**2) * 1000. / sn_gen  # ohm
x_motor = z_motor_abs/np.sqrt(1 + rx**2)
r_motor = rx * x_motor
z_motor = (r_motor + x_motor*1j)/Zbase[1]

#bus currents
i_b1 = np.abs(c / (z_motor + z_line1) + c / z_ext_grid)
i_b1 = np.abs(i_b1 * Ibase[0])

i_b2 = c / (z_ext_grid + z_line1) + c / z_motor
i_b2 = np.abs(i_b2 * Ibase[1])

i_b3 = c / (z_line2 + z_motor * (z_ext_grid + z_line1) / (z_motor + z_ext_grid + z_line1))
i_b3 = np.abs(i_b3 * Ibase[2])

i_bus = np.array([i_b1, i_b2, i_b3])
# array([1.48681907, 0.40324014, 0.21003042])

i_line_00 = np.abs(c / (z_motor + z_line1) * Ibase[0])
i_line_01 = np.abs(c / (z_ext_grid + z_line1) * Ibase[1])
i_line_02 = c / (z_ext_grid + z_line1 + z_motor * z_line2 / (z_motor + z_line2))
i_gen_02 = c / (z_motor + (z_ext_grid + z_line1) * z_line2 / (z_ext_grid + z_line1 + z_line2))
i_line_02 -= i_gen_02 * z_line2 / (z_ext_grid + z_line1 + z_line2)
i_line_02 = np.abs(i_line_02 * Ibase[2])

i_line_10 = 0.
i_line_11 = 0.
i_ext_12 = c / (z_ext_grid + z_line1 + z_motor * z_line2 / (z_motor + z_line2))
i_gen_12  = c / (z_motor +(z_ext_grid + z_line1) * z_line2 / (z_ext_grid + z_line1 + z_line2))
i_line_12 = i_ext_12 * z_motor / (z_motor + z_line2)
i_line_12 += i_gen_12 * (z_ext_grid + z_line1) / (z_ext_grid + z_line1 + z_line2)
i_line_12 = np.abs(i_line_12 * Ibase[2])

i_line_mat = np.array([[i_line_00, i_line_01, i_line_02],
                       [i_line_10, i_line_11, i_line_12]]) 

#array([[0.0447011 , 0.36461306, 0.18991124],
#       [0.        , 0.        , 0.21003042]])

net = motor_three_bus_net()
sc.calc_sc(net, case='min', branch_results=True)
```

Could you share the result that you get from PowerFactory for these two examples? So that I can cross-check.

Reply:
I get these results from PowerFactory:

Bus 1 (Ext Grid Bus)
Ikss = 14.72 (pandapower: 14.43)
Ik = 14.43

Bus 2 (Motor Bus)
Ikss = 6.13  (pandapower: 0.76)
Ik = 0.76

Seems to me like we are not accounting for the difference between Ikss and Ik in motors, and what we are actually calculation is Ik, not Ikss.

Reply:
I would have to check back with the standard about motor modelling. There was also the initiative to seperate motors from static generators in #244 which might be worth revisiting.

Reply:
> I get these results from PowerFactory:
> 
> Bus 1 (Ext Grid Bus)
> Ikss = 14.72 (pandapower: 14.44)
> Ik = 14.43
> 
> Bus 2 (Motor Bus)
> Ikss = 6.13 (pandapower: 0.76)
> Ik = 0.76
> 
> Seems to me like we are not accounting for the difference between Ikss and Ik in motors, and what we are actually calculation is Ik, not Ikss.

Ok, these result seem related to the first example (line 42/51 in `test_motor.py`) that was already there which uses a similar network, only difference is that the ext_grid has `s_sc_max_mva=10.` instead of `s_sc_max_mva=5.`. In the other two examples I added I `s_sc_max_mva` it to 5. in order to magnifies the generator contribution. Is it possible to have PowerFactory results also for these examples?

Reply:
@lthurner 
I tried reproducing the results for the first example:

```
#
# EG---o-----o
#            |
#            G
def motor_net():
    net = pp.create_empty_network()
    b1 = pp.create_bus(net, vn_kv=0.4)
    b2 = pp.create_bus(net, vn_kv=0.4)
    pp.create_ext_grid(net, b1, s_sc_max_mva=10., rx_max=0.1)
    pp.create_line_from_parameters(net, from_bus=b1, to_bus=b2, length_km=1., r_ohm_per_km=0.3211,
                                   x_ohm_per_km=0.06911504, c_nf_per_km=0, max_i_ka=1)
    pp.create_sgen(net, b2, p_mw=0.11, sn_mva=0.5, type="motor", k=7, rx=0.6, current_source=False)
    return net

# Constants
c = 1.1
net = motor_net()
Pbase = net.sn_mva
Ibase = Pbase / (net.bus.vn_kv.values * np.sqrt(3))
Zbase = (net.bus.vn_kv.values**2) / Pbase

# Ext_grid impedance
s_sc_max_mva=10.
rx_max=0.1
z_ext_grid_abs = c/s_sc_max_mva 
x_ext_grid = z_ext_grid_abs/np.sqrt(1+rx_max**2) 
r_ext_grid = x_ext_grid * rx_max 
z_ext_grid = (r_ext_grid + x_ext_grid*1j) * Pbase

#line impedance
z_line =(0.3211 + 0.06911504j)*1.0/Zbase[1]

#motor impedance
k = 7.
sn_gen = 0.5
rx = 0.6
vn_kv = net.bus.loc[net.sgen.bus.values[0]].vn_kv
z_motor_abs = 1./k * (vn_kv**2) / sn_gen #* 1000. # ohm
x_motor = z_motor_abs/np.sqrt(1 + rx**2)
r_motor = rx * x_motor
z_motor = (r_motor + x_motor*1j)/Zbase[1]


#Bus currents
i_b1 = np.abs(c/(z_motor + z_line) + c/z_ext_grid)
i_b1 = i_b1 * Ibase[0]

i_b2 = np.abs(c/(z_ext_grid + z_line) + c/z_motor)
i_b2 = i_b2 * Ibase[1]

i_bus = np.array([i_b1, i_b2])

# With z_motor_abs = 1./k * (vn_kv**2) / sn_gen * 1000.
#array([14.43874584,  0.76396993])

# With z_motor_abs = 1./k * (vn_kv**2) / sn_gen (without multiplying by 1000)
#array([14.72452266,  6.12631924])

# line currents
i_line_00 = np.abs(c/(z_motor + z_line))*Ibase[0]
i_line_01 = np.abs(c/(z_ext_grid + z_line))*Ibase[1]
i_line_mat = np.array([[i_line_00, i_line_01]]) 

# With z_motor_abs = 1./k * (vn_kv**2) / sn_gen * 1000.
#array([[0.00552977, 0.75996391]])
# With z_motor_abs = 1./k * (vn_kv**2) / sn_gen (without multiplying by 1000)
#array([[0.70322661, 0.75996391]])

net = motor_net()
sc.calc_sc(net, case='max', branch_results=True)
```

It appears that the problem is in how the `z_motor` is calculated if I remove the `* 1000.` factor in the formula I get results in line with what you have from powerFactory. Giving a quick look into the code there appear indeed to be a mistake in pandapower implementation (line 581, `build_bus.py`) : 
`z_motor = 1 / (motor.sn_mva.values * 1e-3) / motor.k`
Comparing this with the formula in the docs [here](https://pandapower.readthedocs.io/en/v2.3.0/shortcircuit/voltage_source.html) , this should be changed to:
`z_motor = 1 / (motor.sn_mva.values ) / motor.k`
since the power is in mva and not in kva.